File: Phase3/src/Model/StoreModel.java
package Model;

import DTO.StoreDto;

import java.sql.*;
import java.util.ArrayList;

public class StoreModel {
    private final Oracle database;

    public StoreModel(Oracle database) {
        this.database = database;
    }

    public ArrayList<StoreDto> getStores(ResultSet rs) throws SQLException {
        ArrayList<StoreDto> storeList = new ArrayList<>();
        while (rs.next()) {
            String storeName = rs.getString("Store_Name");
            String address = rs.getString("Address");
            String foodCategory = rs.getString("Food_Category");
            String phoneNumber = rs.getString("Phone_Number");
            String description = rs.getString("Description");
            int deliveryFee = rs.getInt("Delivery_Fee");
            int businessHour = rs.getInt("Business_Hour");
            storeList.add(new StoreDto(
                    storeName,
                    address,
                    foodCategory,
                    phoneNumber,
                    description,
                    deliveryFee,
                    businessHour
            ));
        }
        return storeList;
    }

    public ArrayList<StoreDto> getAllStores() throws SQLException {
        String sql = "SELECT * FROM STORE";
        Statement stmt = this.database.getStatement();
        ResultSet rs = stmt.executeQuery(sql);
        return getStores(rs);
    }

    public ArrayList<StoreDto> getStoresByCategory(String category) throws SQLException {
        String sql = "SELECT * FROM STORE WHERE FOOD_CATEGORY = '" + category + "'";
        Statement stmt = this.database.getStatement();
        ResultSet rs = stmt.executeQuery(sql);
        return getStores(rs);
    }

    public ArrayList<StoreDto> getStoresByMultipleCategory(String category) throws SQLException {
        String[] categorySubStr = category.split("#"); // # 단위로 끊어서 받아오기
        String sql = "";
        for (int i = 0; i < categorySubStr.length; i++) {
            sql += "SELECT * FROM STORE WHERE FOOD_CATEGORY = '" + categorySubStr[i] + "'";
            if (i != categorySubStr.length - 1) {
                sql += " UNION ";
            }
        }
        Statement stmt = this.database.getStatement();
        ResultSet rs = stmt.executeQuery(sql);
        return getStores(rs);
    }

    public ArrayList<StoreDto> getStoresByAddress(String address) throws SQLException {
        String sql = "SELECT * FROM STORE S " +
                "WHERE S.STORE_ID IN (SELECT S1.STORE_ID " +
                "FROM Store S1 " +
                "WHERE S1.ADDRESS LIKE '%" + address + "%')";
        Statement stmt = this.database.getStatement();
        ResultSet rs = stmt.executeQuery(sql);
        return getStores(rs);
    }

    public ArrayList<StoreDto> getStoresByDepartment(int userId) throws SQLException {
        String sql = "SELECT * " +
                "FROM USERS U, DEPARTMENT D, Cooperates C, Store S " +
                "WHERE U.User_ID = ? " +
                "AND U.Dname = C.Dname " +
                "AND C.Store_ID = S.Store_ID " +
                "AND U.Dname = D.Dname";
        PreparedStatement ps = this.database.getPreparedStatement(sql);
        ps.setInt(1, userId);
        ResultSet rs = ps.executeQuery();
        return getStores(rs);
    }

    public StoreDto getStoreByName(String storeName) throws SQLException {
        String sql = "SELECT * FROM STORE WHERE Store_Name = '" + storeName + "'";
        Statement stmt = this.database.getStatement();
        ResultSet rs = stmt.executeQuery(sql);
        return getStores(rs).get(0);
    }

    public String getStoreNameById(int storeId) throws SQLException {
        String storeName = null;
        String sql = "SELECT * FROM STORE WHERE STORE_ID = ?";
        PreparedStatement ps = this.database.getPreparedStatement(sql);
        ps.setInt(1, storeId);
        ResultSet rs = ps.executeQuery();
        if (rs.next()) {
            storeName = rs.getString("Store_Name");
        }
        return storeName;
    }

    public ArrayList<StoreDto> getStoresByMyReview(int userId) throws SQLException {
        String sql = "SELECT DISTINCT * " +
                "FROM Store S, USERS U, Review R " +
                "WHERE U.User_ID = ? " +
                "AND U.User_ID = R.User_ID " +
                "AND R.Store_ID = S.Store_ID";
        PreparedStatement ps = this.database.getPreparedStatement(sql);
        ps.setInt(1, userId);
        return getStores(ps.executeQuery());
    }
}

File: Phase4/server/src/model/StoreModel.ts
import database from "../database";
import { Store } from "../@types/Store";

type StoreDto = {
  STORE_ID: number;
  ADDRESS: string;
  FOOD_CATEGORY: string;
  STORE_NAME: string;
  PHONE_NUMBER: string;
  DESCRIPTION: string;
  DELIVERY_FEE: number;
  IMAGE: string;
  BUSINESS_HOUR: number;
};

class StoreModel {
  getStores = async (sql: string): Promise<Store[] | undefined> => {
    const conn = await database.getConnection();
    const result = (await conn.execute<StoreDto>(sql)).rows;
    return result?.map((store: StoreDto) => {
      return {
        storeId: store.STORE_ID,
        address: store.ADDRESS,
        foodCategory: store.FOOD_CATEGORY,
        storeName: store.STORE_NAME,
        phoneNumber: store.PHONE_NUMBER,
        description: store.DESCRIPTION,
        deliveryFee: store.DELIVERY_FEE,
        image: store.IMAGE,
        businessHour: store.BUSINESS_HOUR,
      };
    });
  };
  getAllStore = async (): Promise<Store[] | undefined> => {
    const sql = "SELECT * FROM STORE";
    return this.getStores(sql);
  };
  getStoreById = async (storeId: number): Promise<Store | undefined> => {
    const sql = `SELECT * FROM STORE WHERE STORE_ID = ${storeId}`;
    const result = await this.getStores(sql);
    return result?.[0];
  };
  getStoresByAddress = async (
    address: string
  ): Promise<Store[] | undefined> => {
    const sql = `SELECT * FROM STORE WHERE ADDRESS = '${address}'`;
    return this.getStores(sql);
  };
  getStoresByFoodCategory = async (
    foodCategory: string
  ): Promise<Store[] | undefined> => {
    const sql = `SELECT * FROM STORE WHERE FOOD_CATEGORY = '${foodCategory}'`;
    return this.getStores(sql);
  };
  getStoresByDepartment = async (
    department: string
  ): Promise<Store[] | undefined> => {
    const sql = `SELECT * FROM STORE WHERE ADDRESS LIKE '%${department}%'`;
    return this.getStores(sql);
  };
}

export default new StoreModel();
